### DISCUSSION TITLE
Features over fixes (for now) ?

### DISCUSSION THREAD

**Discussion by iron3oxide** [reflex-dev/reflex#976] [General]
I know this goes against common wisdom, but I think that at this stage of the project, the focus should not yet be on consolidating/fixing existing features, but on adding new features, namely components/widgets. Let me explain:

As someone who wants to build a beautiful web app in Python without touching JS or CSS, I am very intrigued by this project. The only other frameworks that could potentially satisfy my requirements are [wave](https://wave.h2o.ai/) and [Flet](https://flet.dev/). I think pynecone might "win" in the long term because (amongst other things), it has a team of core developers instead of a single maintainer (Flet) and because state is handled explicitly and obviously instead of being contained in the controls/widgets themselves (both Flet and wave do this to some degree). It also looks somewhat more beautiful by default IMHO.

So why am I still considering the other two? Features, features, features. Some of them I probably won't ever need (PubSub in Flet e.g.), others I definitely will (OAuth in both). My main holdups for each of the frameworks right now are three controls/widgets I will definitely need in my project:

1. A datepicker with great UX (easy to select today or not use the calendar at all) that can be set to different locales
2. A WYSIWYG text editor that can "import" and "export" its content as HTML
3. A datatable that is bound to a model/dataframe with selectable rows

None of the three frameworks has any of these at the time of writing. I have built the datatable for Flet myself from the existing control and would be willing to do so again, the other two can't be glued together in any of the frameworks however and I seem to lack the necessary JS/Dart skills to contribute them myself.

Since I can't wait much longer to start the project I'd need this for, I will probably choose the first of the three to offer these and I'd have to then stick with that framework for a long time, as the project will be somewhat mission critical (bugs are not the end of the world, outages are). For the reasons outlined in the second paragraph, I'd want this to be pynecone. 

Back to the original topic though: I don't think I'm alone with this. There are probably many devs just like me, waiting for 1-3 critical features to be added before they decide to stick with a project. And once they stick, they'll be motivated to fix any bug that might have been introduced by "moving fast and breaking stuff", which I think should be the motto until a stable release is nearing. 

To be clear: I don't want to critique the core devs here, for all I know, they might be following this strategy already or have very good reason not to do so. I rather want to ask them and anyone else reading this : Do you agree with my assessment / see my point? And do you see non-obvious ways to increase the speed at which new features are developed?

**Comment by iron3oxide** [comment]
One proposal I could think of is to let people open an issue for the critical feature they're missing and then a core maintainer adds a complexity label to that issue. If the feature seems to be critical for not a lot of people asides from the person opening the issue, it will only be put on the current sprint once that person has contributed in a manner proportional to the issue complexity, be it by merged PRs, resolved issues or even answered discussions. They will gain code and community insight and probably use the project from thereon out, which makes it more likely they'll keep contributing without the prospect of a feature fast lane.

**Reply by Lendemor** [reply]
I think your suggestion make some good points, but I'll try to nuance them anyway.

While some missing features are definitely needed to engage more people with Pynecone, there are also some bugfixes that are equally needed to make the people that are already adopting Pynecone keep using it.

Some other issues that are actively being worked on involved some refactoring that would be much more time expensive if delayed too long, so getting a solid foundation from the start is better IMHO.

As for your last proposal, anyone can already open any issue they want to suggest additional features or improvement.
We will not ignore the users need, so if any of said issue has a sufficient numbers of people asking for it, it will eventually be implemented.

Last resort, but anyone is free to open a PR if the one feature they need is a niche and unlikely to make it to the tasklist of core maintainers. (On that point, the Discord server is quite active, so any contributors can receive advice if they want to work on a feature but are having some trouble with the existing code base)

Hope it answer some of your concern, let us know what you think about it.

**Reply by iron3oxide** [reply]
I can definitely see your point, right after opening this discussion I opened some issues and PRs regarding bugfixes / DX improvements because I was tasked to provide a nice looking skeleton and thus actually used pynecone for the first time in a realistic context.

#980 actually invigorated me enough to try to do said task in wave instead. I ended up crawling back to pynecone because the architecture of wave isn't very intuitive and doesn't lend itself to the level of modularization I like. But oh did I enjoy those explicit parameters and the accompanying type hints!

Of course I can only speak for myself here, but I feel like the "barrier of entry" is lower for fixing/refactoring existing code than it is for writing completely new code. I know how to fix the issue mentioned above; how to adapt tests and documentation. Creating a new feature on the other hand is a different beast both from a skills and a psychological perspective. I probably need to have more insight into the architecture, coding conventions and testing strategy (maybe even JavaScript or CSS or how the compiler works) than I do for a bug fix/refactoring. But maybe I just need to observe a little bit and it'll all become obvious to me.

Detailed developer documentation might also help with this to some extent. I know from experience though that it's hard to actually capture every implicit assumption/bit of knowledge one has and even more so to keep such documentation up to date. The discord server can be a good way to institutionalize much of that knowledge, I just joined.

**Reply by DanielFrantes** [reply]
Be patient 🙃😉
You have to debug first, then add features.
